### PR TITLE
[SCB-1056] put provider flow control logic in front

### PR DESCRIPTION
**Pull request by yhs0092**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
See details in [SCB-1056](https://issues.apache.org/jira/browse/SCB-1056)

### CODE REVIEW

**Comment by liubao68** [comment]
I don't think change the handler execution sequence is a good idea. I got another solution for using handler logic alone. Here is an example : https://github.com/huaweicse/cse-java-chassis-samples/blob/master/porter/gateway-service/src/main/java/com/huawei/cse/porter/gateway/UiDispatcher.java

I suggest add a configuration, e.g. servicecomb.flowcontrol.transport.enable to make flow controll logic executed before handler, and is a standalone logic, and you can resuse the code of handler. And the execution will not influence any handler logic(that's executed one by one, and if you configured flowcontrol handler, it will also be executed, that's twice execution).

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java [hunk 1]
```diff
@@ -150,6 +167,39 @@ protected void scheduleInvocation() {
...
+          produceProcessor = ProduceProcessorManager.JSON_PROCESSOR;
+          sendResponse(response);
+        });
+      } catch (Exception e) {
+        LOGGER.error("failed to execute ProviderQpsFlowControlHandler", e);
+        qpsFlowControlReject.value = true;
+        sendFailResponse(e);
+      }
+    }
+    return qpsFlowControlReject;
```
[the remark this answers is not in the document]
Comment (yhs0092) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: I don't quite understand this statement.
Do you mean in line 175
```
providerQpsFlowControlHandler.handle(invocation, response -> {
```
the `response` may be null?

**Comment by yhs0092** [comment]
I think a better solution is to refactor the handler mechanism and let the handler execution position more flexiable. The PR just provides a temporary solution.
And I don't use the `ProviderQpsFlowControlHandler` like `UiDispathcer` using `LoadBalanceHandler`, because I want to avoid importing extra maven dependency.
The better solution may require a full refactoring design for handler mechanism. This should be tracked in a new issue.

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java [hunk 2]
```diff
@@ -150,6 +167,39 @@ protected void scheduleInvocation() {
     });
   }
 
+  private Holder<Boolean> checkQpsFlowControl(OperationMeta operationMeta) {
+    Holder<Boolean> qpsFlowControlReject = new Holder<>(false);
+    Handler providerQpsFlowControlHandler = findQpsFlowControlHandler(operationMeta);
+    if (null != providerQpsFlowControlHandler) {
+      try {
+        providerQpsFlowControlHandler.handle(invocation, response -> {
```
[the remark this answers is not in the document]
Comment (liubao68) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: I got it. Do you think it's better to give an option to turn on ProviderQpsFlowControlHandler in AbstractRestInvocation, and do not change the execution flow of handler? That is, when users do want ProviderQpsFlowControlHandler executed in AbstractRestInvocation, they must configure turn the option to true, and remove handler configuration in chain.

**Comment by liubao68** [comment]
Maybe you can make QpsController as common module, and invoke the api in handler & AbstractInvocation. So that low controll in handler logic can remain the same, it's not that bad performance, or we can remove this handler in future.

**Comment by liubao68** [comment]
What do you mean by "a full refactoring design for handler mechanism" ? Optionally put handlers executed in event loop thread?

**Comment by wujimin** [comment]
@liubao68 
i want to redesign mechanism about unify handler and filter and schedule executor, so that developers can arrange their logics in or not in eventloop, not hard code by framework

**Comment by yhs0092** [comment]
@liubao68  As wujimin says, a new handler mechanism should be designed to replace the current handler and filter mechanism. So I think there may be no need to change too much on current code.

**Comment by liubao68** [comment]
Can you create a new issue to tracing the improvement?

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java [hunk 3]
```diff
@@ -150,6 +167,39 @@ protected void scheduleInvocation() {
...
+        LOGGER.error("failed to execute ProviderQpsFlowControlHandler", e);
+        qpsFlowControlReject.value = true;
+        sendFailResponse(e);
+      }
+    }
+    return qpsFlowControlReject;
+  }
+
+  @VisibleForTesting
+  Handler findQpsFlowControlHandler(OperationMeta operationMeta) {
```
Comment (wujimin) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: it's not a good idea to find each time.
we can init and save it in OperationMeta, and mark the new field to be temporary， only for internal usage

**Pull request merged**